### PR TITLE
Change the names for shader compilation settings

### PR DESCRIPTION
**Pull request by JosJuice**
See https://github.com/dolphin-emu/dolphin/pull/8413#issuecomment-544006931

### CODE REVIEW

File: Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp [hunk 1]
```diff
@@ -110,10 +110,10 @@ void GeneralWidget::CreateWidgets()
   auto* shader_compilation_layout = new QGridLayout();
 
   const std::array<const char*, 4> modes = {{
-      QT_TR_NOOP("Synchronous"),
-      QT_TR_NOOP("Synchronous (Ubershaders)"),
-      QT_TR_NOOP("Asynchronous (Ubershaders)"),
-      QT_TR_NOOP("Asynchronous (Skip Drawing)"),
+      QT_TR_NOOP("Traditional"),
```
Comment (blaahaj) [inline Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp]: I like the other names. But I think 'Traditional' is worse than even 'Synchronous'. 'Synchronous' at least tries to explain it, but 'Traditional' only is meaningful if you know what the Tradition is. If it would be changed back, I would fully approve.

Comment (JosJuice) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp]: I guess reverting just that one to "Synchronous" would make sense in a way... It is the only option that compiles shaders synchronously during normal gameplay, after all. What do others think about this?

Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp]: From what we discussed in IRC, basically *everything* is synchronous. Calling one option Synchronous is not accurate, so I'm against that. Traditional works because it is a valid name: it is the traditional method of shader compilation handling in Dolphin and elsewhere.

Comment (JosJuice) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp]: > From what we discussed in IRC, basically *everything* is synchronous.

That was when we were talking about slowdown caused by the GPU performing work. But the other options don't do synchronous **shader compilation**, and shader compilation is exactly what this option is about, which is why I think synchronous makes sense for this option.

Comment (CookiePLMonster) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp]: My first thought is that "Standard" would probably be nice - while it's similar to "Traditional", it is more obvious for the user that they are "normal" or "the most regular".

In an emulator like this, I can imagine people thinking "Traditional" is related to classic/retro modes, as in "looking like it used to on a real console and TV back in the day".

Comment (MayImilae) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp]: "Normal" was the original suggestion, but phire pointed out that it's actually pretty abnormal from a graphics standpooint, it's just the normal for us. Hence, Traditional.

Comment (blaahaj) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp]: It can make sense to an emulator programmer. But to reference assumed knowledge is not a substitute for an attempt at actual description.

**Comment by jordan-woyak** [comment]
This is miles better than the current "Synchronous" naming.
I don't hate "Traditional" but I might suggest "Specialized".

**Comment by stenzek** [comment]
I'm still not a fan of this because it doesn't communicate which shaders are being used, nor how they are being compiled. How does "traditional" mean "specialized shaders and stall emulation until compilation is finished?"

Extending what @jordan-woyak suggested, something like:
 - Specialized
 - Specialized+Ubershaders
 - Ubershaders
 - Specialized+Skip

Would better represent the behavior. At least it specifies which shaders are actually being used, even if the matter in which it affects emulation isn't. But I still prefer the current names.

**Comment by JosJuice** [comment]
@stenzek Something that probably will be a problem with your suggestion is that people are likely to see that there's an option that just says Ubershaders and pick it simply because they've heard that ubershaders are good (even though this option is actually worse than Specialized + Ubershaders). I do think the naming makes sense on a technical level, but that specific thing has the potential to lead to a lot of trouble.

**Comment by nbohr1more** [comment]
I would call it "pre-baked" rather than "specialized" and call Ubershaders "realtime emulated"

**Comment by JosJuice** [comment]
But the specialized shaders aren't pre-baked. They're generated at runtime (barring a shader cache).

**Comment by nbohr1more** [comment]
They are pre-baked in the sense that their code variants are all known a-priori. I guess you could call them "predicted"?

**Comment by JosJuice** [comment]
No, that is exactly what they are not.

**Comment by jordan-woyak** [comment]
@stenzek The problem with the current phrasing is the discrepancy in what's being described.
- A. Synchronous
  - "synchronous" describes waiting on specialized shader compilation.
- B. Synchronous (Ubershaders)
  - "synchronous" describes waiting on Ubershader compilation.
- C. Asynchronous (Ubershaders)
  - "asynchronous" describes not waiting on specialized shader compilation but we still have to wait for Ubershaders which are just as "synchronous" as in mode "B" yet we use the completely opposite word. **This is the problem!**
- D. Asynchronous (Skip Drawing)

The difference from mode "B" to mode "C" is use of specialized shaders after they compile in the background but the current phrasing implies there is less waiting in mode "C" which is not true.
If anything there is more waiting because of poor driver behavior.

"B" describes waiting on the thing in the parentheses, Ubershaders, while "C" describes not waiting on the thing NOT in the parentheses, specialized shaders, which aren't even mentioned. They describe different properties of the implementation and it's confusing. When I want to test the mode that uses just Ubershaders I have to stop and think every time about which of "B" or "C" is the one without specialized shaders.

Every mode is "synchronous" except "D". "C" has some "asynchronous" compilation going on but it is still very much "synchronous".

**Comment by KatherineFtw** [comment]
Ok, so. Let me tackle this from a separate angle. As a user, what do I see for all of these? I see "Ubershaders" and three variants. By it's self, Ubershaders, Traditional, Sync and ASync, Specialized, Skip. Their all meaningless in informing me on what they all do, or very well what they are, mean, or if default is accurate/ok.

At the end of the day, realize that most people do not know, and probably don't care. 

I suggest it's handled like RPCS3's Options section of the wiki. It's helped me more than anything else suggested combined.

https://wiki.rpcs3.net/index.php?title=Help:Default_Settings

TL;DR: No amount of arguing over which jargon to use will actually resolve anything to non-jargon informed users.

**Comment by stenzek** [comment]
@jordan-woyak B/C shouldn't involve any waiting for ubershaders as they are all compiled at startup.

Maybe another way of expressing the options would be better. For example, a slider with specialised shaders on the left, specialised+ubershaders in the middle, ubershaders on the right. So defining performance from fastest to slowest, but stutter from most to least. Skip drawing could be an additional checkbox as it's unrelated to ubershaders, and checking it would disable/grey out the slider.

@KatherineFtw having such documentation is great, but I fear that users won't read it. We're lucky if people read the in-dialog.text, let alone opening their browser and searching :(

**Review by Ebola16 — APPROVED** [review]
As long as the changes are technically correct, I'm ok with the current version of this PR. I had a problem with the previous naming scheme because the names were too similar. People could report a problem with "synchronous" shaders and really mean "Synchronous (Ubershaders)" but they only remembered the first word of the setting.

File: Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp (hunk 1 above)
Comment (blaahaj) [reply Source/Core/DolphinQt/Config/Graphics/GeneralWidget.cpp:116]: Idea: 'Wait For Compile'?

**Comment by MayImilae** [comment]
So pondering this some more, one option for this that we could do is to go with a more accurate name for specialized shaders and just stick (Default) beside it. Like this:

* Specialized (Default)
* Exclusive Ubershaders
* Hybrid Ubershaders
* Skip Drawing

**Comment by JosJuice** [comment]
I would be fine with that. Does anyone have anything against it?

**Comment by mbc07** [comment]
The new names LGTM. 

On a side note, since we're changing the naming, I would probably throw a note on the Exclusive Ubershaders description text to inform the users they should prefer Hybrid Ubershaders instead of Exclusive unless they really know what they're doing. I've seen quite a bunch of people straight up selecting "Exclusive" instead of "Hybrid" thinking it's better then complaining on the forums and on other communities that everything became slow...

**Comment by JosJuice** [comment]
Hm, I wonder what the change should be in more concrete terms... What do you think about replacing "Only recommended for high-end systems" with "Don't use this unless you have a very powerful GPU"?

**Comment by JMC47** [comment]
I'm good with those names.

**Comment by JosJuice** [comment]
I've made the changes.

**Comment by mbc07** [comment]
> What do you think about replacing "Only recommended for high-end systems" with "Don't use this unless you have a very powerful GPU"?

People with high end GPUs might still feel "encouraged" to select this, and that group is also the most likely to be running 6x+ IR with AA and other expensive settings, those coupled with Exclusive Ubershaders can bring even the mightiest of the cards to its knees. 

What about "Requires a very powerful GPU, use only if you still experience shader compilation stuttering with Hybrid Ubershaders"?

**Comment by MayImilae** [comment]
> Hm, I wonder what the change should be in more concrete terms... What do you think about replacing "Only recommended for high-end systems" with "Don't use this unless you have a very powerful GPU"?

I'm ok with this, but I suggest putting in a comment dating it to 2021. What defines a powerful GPU will change and it will be convenient to have a date assigned to it for when it comes time to alter that.

**Comment by JosJuice** [comment]
@mbc07 Would you be fine with "Don't use this unless you have a very powerful GPU and experience shader compilation stuttering with Hybrid Ubershaders"? I feel like "Requires a very powerful GPU" feels a bit redundant with the sentence that comes before.

@MayImilae What do you think would be a good way to include the 2021 wording?

**Comment by JosJuice** [comment]
Ohh, wait, do you mean a source code comment for the 2021 thing? In that case I'll just throw it in there right away.

**Comment by MayImilae** [comment]
Mmmhm, I meant source code comment for the 2021 thing just to give future devs an idea of what "powerful" meant there.

**Comment by mbc07** [comment]
May's description LGTM

**Comment by JosJuice** [comment]
The description is now:
> Ubershaders will always be used. Provides a near stutter-free experience at the cost of very high GPU performance requirements.
>
> **Don't use this unless you encountered stuttering with Hybrid Ubershaders and have a very powerful GPU.**

**Comment by blaahaj** [comment]
3/4 options involve "specialized" shaders. However, only 1/4 options says "specialized"?

**Comment by JMC47** [comment]
This is an improvement over the current situation.  Things we can do in the future is try to better describe what the modes do in the options since they do use mixes of the shaders.  There's no such thing as a Hybrid Ubershader - it's a simplification of a mode that combines ubershaders and specialized shaders.

Anyway, this has languished long enough.  Let's hit the green button and argue about more updates to it in a separate pull request.
